src/main/java/org/openelisglobal/common/rest/provider/NonConformEventsRestController.java

```diff
@@ -0,0 +1,250 @@
+  private final RequesterService requesterService;
+
+  private static final String USER_SESSION_DATA = "userSessionData";
+
+  @Autowired
+  private SystemUserService systemUserService;
+
+  public NonConformEventsRestController(
+    SampleService sampleService,
+    SampleItemService sampleItemService,
+    SearchResultsService searchResultsService,
+    NonConformingEventWorker nonConformingEventWorker,
+    NceCategoryService nceCategoryService,
+    RequesterService requesterService
+  ) {
+    this.sampleService = sampleService;
+    this.sampleItemService = sampleItemService;
+    this.searchResultsService = searchResultsService;
+    this.nonConformingEventWorker = nonConformingEventWorker;
+    this.nceCategoryService = nceCategoryService;
+    this.requesterService = requesterService;
+  }
+
+  @GetMapping(
+    value = "/rest/nonconformevents",
+    produces = MediaType.APPLICATION_JSON_VALUE
+  )
+  public ResponseEntity<?> getNCESampleSearch(
+    @RequestParam(required = false) String lastName,
+    @RequestParam(required = false) String firstName,
+    @RequestParam(required = false) String STNumber,
+    @RequestParam(required = false) String labNumber
+  ) {
+    try {
+      List<Sample> searchResults;
+      if (labNumber != null) {
+        Sample sample = sampleService.getSampleByAccessionNumber(labNumber);
+        searchResults = sample != null ? List.of(sample) : List.of();
+      } else {
+        List<PatientSearchResults> results =
+          searchResultsService.getSearchResults(
+            lastName,
+            firstName,
+            STNumber,
+            "",
+            "",
+            "",
+            "",
+            "",
+            "",
+            ""
+          );
+        searchResults = results
+          .stream()
+          .flatMap(
+            patientSearchResults ->
+              sampleService
+                .getSamplesForPatient(patientSearchResults.getPatientID())
+                .stream()
+          )
+          .collect(Collectors.toList());
+      }
+
+      if (searchResults.isEmpty()) {
+        return ResponseEntity.notFound().build();
+      } else {
+        List<Map<String, Object>> temp = new ArrayList<>();
+        for (Sample sample : searchResults) {
+          temp.add(addSample(sample));
+        }
+        return ResponseEntity.ok().body(Map.of("results", temp));
+      }
+    } catch (Exception e) {
+      return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(
+        "An error occurred while processing the request."
+      );
+    }
+  }
+
+  @GetMapping(
+    value = "/rest/reportnonconformingevent",
+    produces = MediaType.APPLICATION_JSON_VALUE
+  )
+  public ResponseEntity<?> getReportNonConformingEvent(
+    @RequestParam Map<String, String> params,
+    HttpServletRequest request
+  ) {
+    try {
+      Map<String, Object> eventData = new HashMap<>();
+      eventData.put("labOrderNumber", params.get("labOrderNumber"));
```

Comment:
we dont want to work with loosely typed objects , can you Create a Java class representing this Map ??